Draft: Add support for SSE (WagtailVectorIndexSSEConsumer) and async querying (#28)
wagtail/wagtail-vector-index: docs/quick-start.md

```

To skip the prompt, use the `--noinput` flag.

## Using event-stream (WagtailVectorIndexSSEConsumer)

> The WagtailVectorIndexSSEConsumer is a asynchronous HTTP consumer designed for handling Server-Sent Events (SSE), for streaming responses from queries using the vector index in real-time. Using the consumer requires ASGI (uvicorn, daphne etc.) along with configuring django-channels.

Reviewer: ```suggestion
`WagtailVectorIndexSSEConsumer` is an asynchronous HTTP consumer designed for handling Server-Sent Events (SSE) for streaming responses from queries using the vector index in real-time. Using the consumer requires ASGI ([uvicorn](https://pypi.org/project/uvicorn/), [Daphne](https://pypi.org/project/daphne/) etc.) along with [django-channels](https://pypi.org/project/django-channels/).
```

> Configure channels via this [guide](https://channels.readthedocs.io/en/stable/topics/channel_layers.html#configuration), otherwise, simply add channels to INSTALLED_APPS.

Reviewer: ```suggestion
You can configure channels using the [official guide](https://channels.readthedocs.io/en/stable/topics/channel_layers.html#configuration). At a minimum, add `channels` to `INSTALLED_APPS` in your settings file.
```

> ```python

Reviewer: ```suggestion
```python
# settings.py

```

INSTALLED_APPS = [
    "channels",
    # ...
]
```

> You will now need to create an new consumer inheriting WagtailVectorIndexSSEConsumer, but, assigning a page model for the vector index you'd like to use.

Reviewer: ```suggestion
Next, you will need to define a new consumer inheriting from `WagtailVectorIndexSSEConsumer`, and assign a Wagtail page model for the vector index you'd like to use.
```

\*Note: AuthMiddleware is required to provide User context to the consumer.

```python

> import os

Reviewer: suggestion: add a comment with the example path/filename for this

from channels.auth import AuthMiddlewareStack
from channels.routing import ProtocolTypeRouter, URLRouter
from django.core.asgi import get_asgi_application
from django.urls import path, re_path
from wagtail_vector_index.consumers import WagtailVectorIndexSSEConsumer

os.environ.setdefault("DJANGO_SETTINGS_MODULE", "app.settings.production")
django_asgi_app = get_asgi_application()

application = ProtocolTypeRouter(
    {
        "http": URLRouter(
            [
                path(
                    "chat-query-sse/",
                    AuthMiddlewareStack(WagtailVectorIndexSSEConsumer.as_asgi()),
                ),
                re_path(r"", get_asgi_application()),
            ]
        ),
    }
)
```

> You should now be able to query the consumer via the EventSource API, you can use the snippet below as a reference.

Reviewer: ```suggestion
You should now be able to query the consumer using the [EventSource](https://developer.mozilla.org/en-US/docs/Web/API/EventSource) API. The snippet below is an example implementation:
```

```

To skip the prompt, use the `--noinput` flag.

## Using event-stream (WagtailVectorIndexSSEConsumer)

`WagtailVectorIndexSSEConsumer` is an asynchronous HTTP consumer designed for handling Server-Sent Events (SSE) for streaming responses from queries using the vector index in real-time. Using the consumer requires ASGI ([uvicorn](https://pypi.org/project/uvicorn/), [Daphne](https://pypi.org/project/daphne/) etc.) along with [django-channels](https://pypi.org/project/django-channels/).

> You can configure channels using the [official guide](https://channels.readthedocs.io/en/stable/topics/channel_layers.html#configuration). At a minimum, install the `channels` package and add it to `INSTALLED_APPS` in your settings file.

Reviewer 2: **Question:** I'm imagining that given we're not using channel layers, we don't require Redis to run channels, is that correct? Do we actually need to set up channel layers at all? Is this configuration requirement a bit misleading?

Author: Correct, the channels layer isn't a dependency, however, the channels package itself is, channels is required for the AsyncHttpConsumer, although now I'm writing this I've noticed that you're talking about the link, which points to the channels layer. 

Fixed by updating the link.

```python
# settings.py

INSTALLED_APPS = [
    "channels",
    # ...
]
```

> Next, you will need to define a new consumer inheriting from `WagtailVectorIndexSSEConsumer`, and assign a Wagtail page model for the vector index you'd like to use.

Reviewer 3: **Issue:** I believe this is no longer true now we use a single consumer across all indexes?

```

To skip the prompt, use the `--noinput` flag.

> ## Using event-stream (WagtailVectorIndexSSEConsumer)

Author: Not quite sure this belongs in the "quick-start" section, might be better off in it's own markdown file

The WagtailVectorIndexSSEConsumer is a asynchronous HTTP consumer designed for handling Server-Sent Events (SSE), for streaming responses from queries using the vector index in real-time. Using the consumer requires ASGI (uvicorn, daphne etc.) along with configuring django-channels.

Configure channels via this [guide](https://channels.readthedocs.io/en/stable/topics/channel_layers.html#configuration), otherwise, simply add channels to INSTALLED_APPS.

```python
INSTALLED_APPS = [
    "channels",
    # ...
]
```

> and set the channel layer. The [in-memory channel layer](https://channels.readthedocs.io/en/stable/topics/channel_layers.html#in-memory-channel-layer) can be used for local envrionments, however, redis is the only offically maintained channel layer supported for production use, thus [channels_redis](https://pypi.org/project/channels-redis/) will have to be install.

Reviewer 3: **question:** Do we need to use channel layers? I can't see them used in the consumer

Author: Removed, Thanks for spotting this one. I think I added channels in when experimenting with communicating to the consumer from a sync view and assumed it has been a requirement since.

```
CHANNEL_LAYERS = {
    "default": {
        "BACKEND": "channels_redis.core.RedisChannelLayer",
        "CONFIG": {
            "hosts": [("127.0.0.1", 6379)],
        },
    },
}
```

You will now need to create an new consumer inheriting WagtailVectorIndexSSEConsumer, but, assigning a page model for the vector index you'd like to use.

*Note: AuthMiddleware is required to provide User context to the consumer.

```python
import os

from channels.auth import AuthMiddlewareStack
from channels.routing import ProtocolTypeRouter, URLRouter
from django.core.asgi import get_asgi_application
from django.urls import path, re_path
from wagtail_vector_index.consumers import WagtailVectorIndexSSEConsumer

os.environ.setdefault("DJANGO_SETTINGS_MODULE", "app.settings.production")
django_asgi_app = get_asgi_application()

class WikiPageChatQuerySSEConsumer(WagtailVectorIndexSSEConsumer):
    page_model_name = "wiki.WikiPage"

application = ProtocolTypeRouter(
    {
        "http": URLRouter(
            [
                path(

>                     "chat-query-sse/<uuid:channel_uuid>/",

Reviewer 3: **question**: What is the UUID for?

Author: Same as the reply to the channel layer question https://github.com/wagtail/wagtail-vector-index/pull/28#discussion_r1434131184, left in from some previous code I didn't consider removing.

]
```

You will now need to create an new consumer inheriting WagtailVectorIndexSSEConsumer, but, assigning a page model for the vector index you'd like to use.

> \*Note: AuthMiddleware is required to provide User context to the consumer.

Reviewer: ```suggestion
!!! Note
    The `AuthMiddleware` is required to provide the user context to the consumer.
```

### Known issues

> Asynchronous support in Django is fairly new, as a result of this WagtailVectorIndexSSEConsumer can't tell when a client disconnects from an event-stream, This may result in queries being processed by the server as zombie threads.

Reviewer: ```suggestion
Asynchronous support in Django is fairly new and `WagtailVectorIndexSSEConsumer` can't tell when a client disconnects from an event-stream. This may result in queries being processed by the server as zombie threads.
```
